### PR TITLE
BigDecimal field values must be rounded according to field format #968

### PR DESCRIPTION
**Pull request by Gavrilov-Ivan**
https://github.com/jmix-framework/jmix/issues/968
Pay close attention to property - is it ok to put it in 'jmix.core'?

### CODE REVIEW

**Comment by gorbunkov** [comment]
@Gavrilov-Ivan Did you check whether this problem is also relevant for Double or Float datatypes?

**Comment by Gavrilov-Ivan** [comment]
Yes, it's relevant for Double and Float datatypes also.
Modified parse process of them the similar way:

1. Parse raw string value as BigDecimal
2. Perform rounding if necessary
3. Check if parsed value fits value-range of target datatype.
4. Get result value as a target datatype

Added tests.

File: jmix-core/core/src/test/groovy/properties/TestCoreProperties.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+ */
+
+package properties;
+
+import io.jmix.core.CoreProperties;
+
+import java.time.Duration;
+import java.util.List;
+
+public class TestCoreProperties extends CoreProperties {
```
Comment (gorbunkov) [inline jmix-core/core/src/test/groovy/properties/TestCoreProperties.java:24]: It seems that "groovy" directory is a wrong place for java class. `java/test_support/datatypes` seems to be a better choice.
Also, what is the purpose of the class? Is it for setting the `decimalValueRoundByFormat` property only? If so, then maybe we could add the `@TestPropertySource` annotations to test classes with explicit property value instead?

File: jmix-core/core/src/main/java/io/jmix/core/metamodel/datatype/impl/DoubleDatatype.java [hunk 2]
```diff
@@ -82,10 +90,39 @@ public Double parse(String value, Locale locale) throws ParseException {
...
+    protected Number parse(String value, NumberFormat format) throws ParseException {
+        BigDecimal result = (BigDecimal) super.parse(value, format);
+        if (coreProperties.isDecimalValueRoundByFormat()) {
+            int maximumFractionDigits = format.getMaximumFractionDigits();
+            RoundingMode roundingMode = format.getRoundingMode();
+            result = result.setScale(maximumFractionDigits, roundingMode);
+        }
+
+        if (!isInDoubleRange(result)) {
+            throw new ParseException(String.format("Double range exceeded: \"%s\"", value), 0);
```
Comment (gorbunkov) [inline jmix-core/core/src/main/java/io/jmix/core/metamodel/datatype/impl/DoubleDatatype.java:107]: I'm not sure that "exceeded" is a proper word if we speak about negative values. Maybe we can say something like "The value is out of Double datatype range"? The same for Float.

File: jmix-core/core/src/main/java/io/jmix/core/CoreProperties.java [hunk 3]
```diff
@@ -100,6 +100,12 @@ public class CoreProperties {
      */
     PessimisticLock pessimisticLock;
 
+    /**
+     * Whether BigDecimalDatatype, DoubleDatatype and FloatDatatype should round the actual parsed value
+     * according to format settings.
+     */
+    boolean decimalValueRoundByFormat;
```
Comment (gorbunkov) [inline jmix-core/core/src/main/java/io/jmix/core/CoreProperties.java:107]: Maybe rename the property to `roundDecimalValueByFormat`? WDYT?

**Review by gorbunkov — CHANGES_REQUESTED** [review]

File: jmix-core/core/src/main/java/io/jmix/core/CoreProperties.java (hunk 3 above)
Comment (Gavrilov-Ivan) [reply jmix-core/core/src/main/java/io/jmix/core/CoreProperties.java]: Looks a little better - renamed

File: jmix-core/core/src/main/java/io/jmix/core/metamodel/datatype/impl/DoubleDatatype.java (hunk 2 above)
Comment (Gavrilov-Ivan) [reply jmix-core/core/src/main/java/io/jmix/core/metamodel/datatype/impl/DoubleDatatype.java]: Done

File: jmix-core/core/src/test/groovy/properties/TestCoreProperties.java (hunk 1 above)
Comment (Gavrilov-Ivan) [reply jmix-core/core/src/test/groovy/properties/TestCoreProperties.java]: Moved class to `java/test_support`, without `datatypes` sub-package - this class can be used not only for `datatype` scope of tests.
This class helps to change property values in different tests within single test class. Adding a 'static' property source for entire class looks less suitable for this case. Don't want to split it into multiple test classes based on possible values of involved properties.

File: jmix-core/core/src/main/java/io/jmix/core/metamodel/datatype/impl/BigDecimalDatatype.java [hunk 4]
```diff
@@ -94,7 +98,13 @@ public BigDecimal parse(String value, Locale locale) throws ParseException {
         DecimalFormatSymbols formatSymbols = formatStrings.getFormatSymbols();
         DecimalFormat format = new DecimalFormat(formatStrings.getDecimalFormat(), formatSymbols);
         format.setParseBigDecimal(true);
-        return (BigDecimal) parse(value, format);
+        BigDecimal result = (BigDecimal) parse(value, format);
+        if(coreProperties.isBigDecimalValueRoundByFormat()) {
```
[the remark this answers is not in the document]
Comment (Gavrilov-Ivan) [reply jmix-core/core/src/main/java/io/jmix/core/metamodel/datatype/impl/BigDecimalDatatype.java]: Had been done earlier

**Review by gorbunkov — APPROVED** [review]

**Review by glebfox — APPROVED** [review]

**Pull request merged**